crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -355,6 +572,83 @@ impl<'a> LimitsTraversal<'a> {
     }
 }
 
+impl<'a> Reporter<'a> {
+    fn new(ctx: &'a ExtensionContext<'a>) -> Self {
+        let cfg: &ServiceConfig = ctx.data_unchecked();
+        Self {
+            limits: &cfg.limits,
+            query_id: ctx.data_unchecked(),
+            session_id: ctx.data_unchecked(),
+        }
+    }
+
+    /// Error returned if a fragment is referred to but not found in the document.
+    fn fragment_not_found_error(&self, name: &Name, pos: Pos) -> ServerError {
+        self.graphql_error_at_pos(
+            code::BAD_USER_INPUT,
+            format!("Fragment {name} referred to but not found in document"),
+            pos,
+        )
+    }
+
+    /// Error returned if output node estimate exceeds limit.
+    fn output_node_error(&self) -> ServerError {
+        self.graphql_error(
+            code::BAD_USER_INPUT,
+            format!(
+                "Estimated output nodes exceeds {}",
+                self.limits.max_output_nodes
+            ),
+        )
+    }
+
+    /// Error returned if the payload size exceeds the limit.
+    fn payload_size_error(&self, message: &str) -> ServerError {
+        self.graphql_error(
+            code::BAD_USER_INPUT,
+            format!(
+                "{message}. Requests must spend {max_tx_payload} bytes or fewer on transaction \
```

Comment:
```suggestion
                "{message}. Requests are limited to {max_tx_payload} bytes or fewer on transaction \
```

Looking through the test examples in `builder.rs` I was a bit confused by the "must spend" language